lib/spack/spack/tag.py

```diff
@@ -0,0 +1,136 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+"""Classes and functions to manage package tags"""
+import collections
+import copy
+import sys
+
+if sys.version_info >= (3, 5):
+    from collections.abc import Mapping  # novm
+else:
+    from collections import Mapping
+
+import spack.error
+import spack.util.spack_json as sjson
+
+
+def get_installed_packages():
+    """Returns names of packages installed in the active environment."""
+    env = spack.environment.active_environment()
+    hashes = env.all_hashes() if env else None
+    return [spec.name for spec in spack.store.db.query(hashes=hashes)]
+
+
+def get_tag_packages(tags, installed, skip_empty):
+    """
+    Returns a dict, indexed by tag, containing lists of names of packages
+    containing the tag or, if no tags, for all available tags.
+
+    Arguments:
+        tags (list or None): list of tags of interest or None for all
+        installed (bool): True if want names of packages that are installed;
+            otherwise, False if want all packages with the tag
+        skip_empty (bool): True if exclude tags with no associated packages;
+            otherwise, False if want entries for all tags even when no such
+            tagged packages
+    """
+    tag_pkgs = collections.defaultdict(lambda: list)
+    spec_names = get_installed_packages() if installed else []
+    keys = tags or spack.repo.path.tag_index
```

Comment:
Should we be more defensive here:
```suggestion
    keys = spack.repo.path.tag_index if tags is None else tags
```
? Asking this as a question since on the one hand I like the `or` idiom but on the other I can see the API behaving the opposite of what would be expected if we happen to call this function with an empty list. Think for instance if the list is constructed during a script and ends up being empty.

Reply:
🤔

Reply:
The case I have in mind here is:
```python
# Assume this is the result of a computation and is empty
tags = []
result = get_tag_packages(tags, installed=False, skip_empty=True)
```
`result` will contain all the tags, while I assume the client expects it to be empty. It is also inconsistent with the documentation:
```
tags (list or None): list of tags of interest or None for all
```
because `None` is different from the empty list (but still false-ish)

Reply:
You're absolutely right.  (I was side-tracked by something else so am now getting back to this.)